src/index/PatternIndex.cpp

```diff
@@ -0,0 +1,834 @@
+  // into the file.
+  constexpr size_t CHUNK_SIZE = 2048;
+  char chunk[CHUNK_SIZE * sizeof(uint64_t) * sizeof(uint32_t)];
+  size_t num_chunks = entityHasPattern.size() / CHUNK_SIZE;
+  if (entityHasPattern.size() % CHUNK_SIZE != 0) {
+    num_chunks++;
+  }
+  std::cout << "num chunks " << num_chunks << std::endl;
+  for (size_t chunk_idx = 0; chunk_idx < num_chunks; chunk_idx++) {
+    size_t offset = chunk_idx * CHUNK_SIZE;
+    size_t count = std::min(CHUNK_SIZE, entityHasPattern.size() - offset);
+    // Copy the data skipping the struct padding
+    for (size_t i = 0; i < count; ++i) {
+      std::memcpy(chunk + (i * 12), &entityHasPattern[offset + i].subject, 12);
+    }
+    // write a chunk to the file
+    file.write(chunk, count * (sizeof(uint64_t) + sizeof(uint32_t)));
+  }
+
+  // write the entityHasPredicate vector
+  size_t numHasPredicates = entityHasPredicate.size();
+  std::cout << "hasPredicatesize" << numHasPredicates << std::endl;
+  file.write(&numHasPredicates, sizeof(size_t));
+  static_assert(sizeof(std::array<Id, 2>) == sizeof(Id) * 2,
+                "std::array<Id, 2> has padding, unable to serialize and "
+                "deserialize the hasPredicate vector.");
+  file.write(entityHasPredicate.data(), sizeof(Id) * numHasPredicates * 2);
+
+  // write the patterns
+  pattern_data->patterns().write(file);
+  file.close();
+
+  LOG(INFO) << "Done creating patterns file." << std::endl;
+
+  // create the has-relation and has-pattern lookup vectors
+  if (entityHasPattern.size() > 0) {
+    std::vector<PatternID>& hasPattern = pattern_data->hasPattern();
+    hasPattern.resize(entityHasPattern.back().subject + 1);
+    size_t pos = 0;
+    for (size_t i = 0; i < entityHasPattern.size(); i++) {
+      while (entityHasPattern[i].subject > pos) {
+        hasPattern[pos] = NO_PATTERN;
+        pos++;
+      }
+      hasPattern[pos] = entityHasPattern[i].pattern;
+      pos++;
+    }
+  }
+
+  vector<vector<PredicateId>> hasPredicateTmp;
+  if (entityHasPredicate.size() > 0) {
+    hasPredicateTmp.resize(entityHasPredicate.back()[0] + 1);
+    size_t pos = 0;
+    for (size_t i = 0; i < entityHasPredicate.size(); i++) {
+      Id current = entityHasPredicate[i][0];
+      while (current > pos) {
+        pos++;
+      }
+      while (i < entityHasPredicate.size() &&
+             entityHasPredicate[i][0] == current) {
+        hasPredicateTmp.back().push_back(entityHasPredicate[i][1]);
+        i++;
+      }
+      pos++;
+    }
+  }
+  pattern_data->hasPredicate().build(hasPredicateTmp);
+}
+
+// _____________________________________________________________________________
+void PatternIndex::throwExceptionIfNotInitialized() const {
+  if (!_initialized) {
+    AD_THROW(ad_semsearch::Exception::CHECK_FAILED,
+             "The requested feature requires a loaded patterns file ("
+             "do not specify the --no-patterns option for this to work)");
+  }
+}
```

Comment:
I only skimmed this last part, I assume it is mostly a copy of the original pattern creation with adapted templated types.
In general we should use a proper symmetric serialization mechanism in the future to avoid mistakes with the manual read and write calls.